File: legacy/isma-tools/src/main/java/ru/nstu/isma/generate/MemoryFileManager.java
package ru.nstu.isma.generate;

/**
 * Created by Bessonov Alex
 * on 10.12.2014.
 */

import javax.tools.FileObject;
import javax.tools.ForwardingJavaFileManager;
import javax.tools.JavaFileObject;
import javax.tools.JavaFileObject.Kind;
import javax.tools.StandardJavaFileManager;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.security.SecureClassLoader;

public class MemoryFileManager extends ForwardingJavaFileManager<StandardJavaFileManager> {

    private MemoryJavaClassObject object;

    public MemoryFileManager(StandardJavaFileManager manager) {
        super(manager);
    }

    @Override
    public ClassLoader getClassLoader(Location location) {
        return new SecureClassLoader() {
            @Override
            protected Class<?> findClass(String name) throws ClassNotFoundException {
                byte[] b = object.getBytes();
                return super.defineClass(name, object.getBytes(), 0, b.length);
            }

            @Override
            public InputStream getResourceAsStream(String name) {
                if (object.getName().endsWith(name)) {
                    return new ByteArrayInputStream(object.getBytes());
                }
                return super.getResourceAsStream(name);
            }
        };
    }

    @Override
    public JavaFileObject getJavaFileForOutput(Location location, String name, Kind kind, FileObject sibling) throws IOException {
        object = new MemoryJavaClassObject(name, kind);
        return object;
    }

}

File: isma-compiler/jvm-utils/src/main/kotlin/ru/nstu/isma/generate/MemoryFileManager.kt
package ru.nstu.isma.generate

import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.security.SecureClassLoader
import javax.tools.*

/**
 * Created by Bessonov Alex
 * on 10.12.2014.
 */
class MemoryFileManager(manager: StandardJavaFileManager?) :
    ForwardingJavaFileManager<StandardJavaFileManager?>(manager) {
    private var obj: MemoryJavaClassObject? = null
    override fun getClassLoader(location: JavaFileManager.Location?): ClassLoader {
        return object : SecureClassLoader() {
            @Throws(ClassNotFoundException::class)
            override fun findClass(name: String): Class<*> {
                val b = obj!!.bytes
                return super.defineClass(name, b, 0, b.size)
            }

            override fun getResourceAsStream(name: String): InputStream? {
                return if (obj!!.name.endsWith(name)) {
                    ByteArrayInputStream(obj!!.bytes)
                } else super.getResourceAsStream(name)
            }
        }
    }

    @Throws(IOException::class)
    override fun getJavaFileForOutput(
        location: JavaFileManager.Location,
        name: String,
        kind: JavaFileObject.Kind,
        sibling: FileObject
    ): JavaFileObject {
        obj = MemoryJavaClassObject(name, kind)
        return obj!!
    }
}